File: python3源码/Test05.py
# -*- coding:utf-8 -*-

# 剑指offer-题5：从尾到头打印链表

# 输入一个链表，从尾到头打印链表每个节点的值。


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class digui_Solution:  # 递归
    value_list = []
    def printListFromTailToHead(self, listNode):
        if listNode is not None:
            self.printListFromTailToHead(listNode.next)
            self.value_list.append(listNode.val)
        return self.value_list


class Solution:  # 栈
    def printListFromTailToHead(self, listNode):
        stack = []
        value_list = []
        node = listNode
        while node is not None:
            stack.append(node)
            node = node.next
        while len(stack) > 0:
            value_list.append(stack[-1].val)
            stack.pop()
        return value_list


a = ListNode(5)
b = ListNode(8)
c = ListNode(61)
d = ListNode(95)
a.next = b
b.next = c
c.next = d

solu = Solution()
list = solu.printListFromTailToHead(a)
print(list)
solu2 = digui_Solution()
list2 = solu2.printListFromTailToHead(a)
print(list2)

File: java源码/offer/Test05.java
package offer;

/**
 * 剑指offer-题5：从尾到头打印链表
 * 
 * 输入一个链表，从尾到头打印链表每个节点的值。
 */

import java.util.ArrayList;
import java.util.Stack;

public class Test05 {
	ArrayList<Integer> valueList = new ArrayList<Integer>();

	public ArrayList<Integer> printListFromTailToHead(ListNode listNode) {
		/*
		 * if(listNode != null){ //利用递归 printListFromTailToHead(listNode.next);
		 * valueList.add(listNode.val); }
		 */
		Stack<Integer> valueStack = new Stack<Integer>(); // 利用栈
		ListNode node = listNode;
		while (node != null) {
			valueStack.push(node.val);
			node = node.next;
		}
		while (valueStack.size() > 0) {
			valueList.add(valueStack.peek());
			valueStack.pop();
		}

		return valueList;
	}

	public static void main(String[] args) {
		ListNode a = new ListNode(5);
		ListNode b = new ListNode(8);
		ListNode c = new ListNode(61);
		ListNode d = new ListNode(95);
		a.next = b;
		b.next = c;
		c.next = d;

		ArrayList<Integer> valueList = new Test05().printListFromTailToHead(a);
		for (Integer integer : valueList) {
			System.out.print(integer + ", ");
		}
	}

}
